validmind/models/rag.py

```diff
@@ -0,0 +1,112 @@
+# Copyright © 2023-2024 ValidMind Inc. All rights reserved.
+# See the LICENSE file in the root of this repository for details.
+# SPDX-License-Identifier: AGPL-3.0 AND ValidMind Commercial
+
+from dataclasses import dataclass
+from typing import Union
+
+import pandas as pd
+
+from validmind.logging import get_logger
+from validmind.models.foundation import Prompt
+from validmind.models.functional import FunctionalModel
+from validmind.vm_models.model import VMModel
+
+logger = get_logger(__name__)
+
+
+@dataclass
+class RAGPrompt(Prompt):
+    def __post_init__(self):
+        if "context" not in self.variables:
+            raise ValueError("Prompt must contain 'context' variable")
+
+    template: str
+    variables: list
+
+
+DEFAULT_PROMPT_TEMPLATE = """
+Please generate a response to the following query:
+```
+{x}
+```
+
+Using the information provided in the following context:
+```
+{context}
+```
+""".strip()
+
+DEFAULT_PROMPT = RAGPrompt(
+    template=DEFAULT_PROMPT_TEMPLATE,
+    variables=["x", "context"],
+)
+
+
+class EmbeddingModel(FunctionalModel):
```

Comment:
Currently, this is configured to output embeddings to a single column specified by `output_column`. In the case where users want to compute embeddings across many columns, is the preferred approach to call the model separately for each column we want to process, or should the model support generating embeddings for multiple columns in a single call?

Reply:
this is a really good point @juanmleng... I think maybe this needs to be a discussion. I think this goes for other types of models as well since the dataset class expects predictions to be a single column 🤔

Reply:
what i posted in the slack dm should take care of this case